### DISCUSSION TITLE
Stop rounding up individual jobs to the nearest minute for billing

### DISCUSSION THREAD

**Discussion by splitice** [community/community#8726] [Actions] [answered]
A good well structured deployment script in Github Actions can involve many quick jobs. In our case we are aproaching 16 (typical execution time of 4-6 seconds), it's a bit unreasonable to be paying for 16 minutes per job for a total of just over 1 minute of CPU time.

This encourages us to create larger monolithic worflows rather than clean and re-usable ones. If per minute rounding was instead applied on the entire workflow this would encourage correct usage and enable more use cases.

I understand the need to make money, just wish you could be fairer with the distribution of costs.

**Comment by ethomson** [comment]
Hi @splitice, I understand where you're coming from - certainly the current setup for GitHub Actions workflows is optimized for longer-running jobs.  But I'm not sure that multiple _jobs_ is more "correct" unless you actually need a separate virtual machine for each step.

I'd love to know a bit more about what a well-structured deployment script looks like to you...  is this something that you're willing to share?  If not, I'd love to get the highlights - what are these jobs doing?  Philosophically, why do you think of them as separate _jobs_ instead of separate _steps_ within a job?  How are you sharing data between them (or are you not)?  Are you using jobs as an isolation environment?  If so, could you use containers as that isolation environment or do you require really _strict_ isolation?  Is it about the parallelization of jobs?

Regarding the costs, it's important to remember that each job runs on a separate virtual machine, and each new virtual machine takes time to set up and tear down.  We keep runners in the set-up state so that your jobs can start on them as quickly as possible.  We only bill for the time that your job is actually executing on our runners, and the remainder of the time we retain as costs.

**Comment by splitice** [comment]
@ethomson I can't share the workflow of the job that inspired this however I can share s screenshot of a similar one. It's a CD job for an In Development application (and it's already at 12 jobs - expected to be over 30 when complete).

![chrome_2021-12-15_09-51-02](https://user-images.githubusercontent.com/468579/146092269-92633f2f-ff1a-4fcd-8833-30c1837a2a14.png)

Could I make these into a single job? Probably but not without sacrificing alot of the provided functionality. For example:
 - Would it take longer per deploy (likely with sequential execution of dependencies or stages)? certainly
 - Would it be more monolithic and less portable? certainly
 - Would it be more difficult to troubleshoot, less visual and require more engineer effort to understand? certainly

I understand you have to spin up / spin down runners (unfortunately for workloads like this an overkill) but it doesnt change my opinion regarding how billing rounding should be implemented. A change like this would allow people to really put the runner side ("if", matrix builds, etc) to work for lightweight CD jobs rather than encouraging designing to minimising cost (expecially long term as the workflow grows).

To directly answer your questions:

> what are these jobs doing? 

Continuous Deployment for staging / production environments. Sometimes to one target, sometimes we do matrix builds on top of these jobs as well (e.g to deploy to multiple datacenters).

> Philosophically, why do you think of them as separate jobs instead of separate steps within a job? 

Non-linear dependencies, different tooling requirements and (in some cases) sometimes different destination clusters (e.g some services are k8s, some bare metal or not software related).

> How are you sharing data between them (or are you not)? 

Not applicable for this example each step is stateless.

> Are you using jobs as an isolation environment? 

Not particularly. Without isolation some of the jobs would be 1-2s faster (as most jobs have a small amount of overhead for initialization of tooling / environment) however that's "fast enough" as is.

> If so, could you use containers as that isolation environment or do you require really strict isolation?

Certainly containers would be more than sufficient

> Is it about the parallelization of jobs?

Parallel execution when conditions are met is necessary to ensure timely execution of complex jobs. Although these jobs require little CPU (at-least when compared to serious unit tests) they do require authentication and communication with remote services with the latencies inherent with such actions (e.g comparing of service versions or pulling a docker image). 

In situations like this linear execution of jobs would be significantly slower and less optimal. The example I provided would take ~2:30min  vs ~1 minute at it's current size.

**Reply by jsoref** [reply]
I'm not sure if you've seen, but GitHub now has parallel steps (workflow level only; not actions): https://github.blog/changelog/2026-06-25-actions-steps-can-now-be-run-in-parallel/

based on an informal read of your requirements, I suspect that you could use them to at least collapse the first two columns. With some effort, you could probably collapse all four columns into a single job.

- See the original discussion for it https://github.com/community/community/discussions/14484

I'm pretty sure they'd love to hear your experience if you try to use them.

**Comment by ethomson** [comment]
I really appreciate your detailed feedback.  This is really useful to understand for our planning.

Having a runner for each of your jobs is indeed overkill in this scenario.  Right now a job is affined to a runner, and a runner is expensive, which doesn't really fit in well with what you're trying to do here.  We've talked a bit about how to support these sorts of lighter-weight jobs, and but we don't have anything to announce yet.

**Reply by maxisam** [reply]
Can we at least use 30s instead of a minute? It doesn't address the issue completely but at least it is not as bad.

**Reply by jsoref** [reply]
Fwiw, they did eventually release `ubuntu-slim` which is ~1/3 the cost of an ubuntu runner (and only has one cpu instead of 2):
https://docs.github.com/en/actions/reference/runners/github-hosted-runners#single-cpu-runners
https://docs.github.com/en/billing/concepts/product-billing/github-actions#baseline-minute-costs

**Comment by bencodner** [comment]
I am having the same issue. I created a matrix job in a pipeline that packages umbrella helm charts (around 90). Each matrix takes roughly 30 seconds but I'll be changing this to a script because this matrix approach is using too many minutes.

**Comment by nhuray** [comment]
I might understand that Github actions billable time was rounded to the minute 3 years ago when the service was still young, but 3 years later I just don't understand why Github does not charge a fair price based on real usage. 

AWS for instance is now charging to the **milliseconds** for Lambda. I understand it's not the same architecture but paying 10 minutes for 10 jobs running 5 sec seems to me a non-sense.

**Comment by shealavington** [comment]
I've just experienced a very similar thing — Now moving into 2023 🤦‍♂️

I run two jobs, one for Javascript Linting, and one for PHP Linting... In total, it performs on average 30 seconds of work between them, but I'm charged 2 minutes every time 👀 It's on the small scale, but I can see this piling up over time. That's a time charge of 4x my actual usage.

Seems silly for me to crunch this into a single Linting job, because then I also can't keep track of which job/task is taking longer if runtime was to start to spike and they also have different requirements to run.

![image](https://user-images.githubusercontent.com/16869302/220992963-9786a80a-327e-471a-a6c1-26551a41f7ce.png)

Below is another part of my workflow, it's a task for compiling assets, ready to cache the data for subsequent jobs that may or may not need to run. The Testing jobs run parallel because, well, they're testing jobs and either may or may not be needed. 

![image](https://user-images.githubusercontent.com/16869302/221131894-668c5dda-70b9-42d3-b716-76460c117fd2.png)

In this setup I'm charged 3 minutes for less than a minutes work! Excuse me too for trying to be efficient and only having my tests run when they're needed. Instead of this nice workflow, I'll now be forced to always run at least my unit test suite, and design the job to handle caching of assets for any other jobs. At least then I can shave a guaranteed minute off every time.

It's just not as clean, I suppose I'll save 7 seconds doing so, but I'm forced to always run my tests even if not needed now.

**Surely Github should have an efficient way of containerizing for a few seconds and only charging me for actual usage by now?**

**Comment by dataplor-mike** [comment]
Came across this thread while trying to figure out how 11s resulted in a 54x charge of 10 minutes.  Not sure if it's related to rounding or if it's a bug that overcharges for reruns or unmet conditions:

<img width="962" alt="Screenshot 2023-03-17 at 3 43 43 PM" src="https://user-images.githubusercontent.com/113470104/226060091-eb1c417f-fce3-4bfa-bfb1-d441d9e6de0a.png">

**Comment by james-callahan** [comment]
We hit this issue fairly hard, e.g.

![image](https://user-images.githubusercontent.com/35791147/227424010-ab9156ce-71fa-492f-ab31-9e719fd080b7.png)

![image](https://user-images.githubusercontent.com/35791147/227423932-9d0b2aca-5ab8-4e4e-ad43-14dad8303023.png)

--> in many circumstances we're getting charged over 60x more than we would be otherwise.

> I'd love to know a bit more about what a well-structured deployment script looks like to you... is this something that you're willing to share? If not, I'd love to get the highlights - what are these jobs doing? Philosophically, why do you think of them as separate jobs instead of separate steps within a job?

@ethomson in our case we have many jobs as they are generated from a matrix (one for each folder in the repository).

**Comment by thomasschaller1** [comment]
We hit the same issue. We are evaluating different solutions, and this issue does not help using concurrency. Because instead of just building something, and then running multiple job with the result of the build, we will have to put everything in the same job to reduce our costs, and this will double the time of the whole pipeline (1min vs 2min).

In 2023, we can expect to have something at the seconds level, when you see that AWS is doing it even in the milliseconds level.

**Comment by sigi-tw** [comment]
How 'un'funny: just splitted up my workflow in a pipeline with sub jobs (github docu also says its a good way to reuse workflows) and now i'm discovering this specific issue.

The main reason why i splitted it up was because i have two images which need to be build and successful and then i need to do an e2e test.

**Comment by dhouglasoliveiraatpn** [comment]
Started on use of matrix stragegy to reuse jobs, but with this in mind, is a NO-GO!
![image](https://github.com/community/community/assets/102536045/25540597-5342-4f7a-acbe-6aa8f177162b)

**Reply by ollytheninja** [reply]
Reason number 1 why GitHub likes matrix jobs so much!

**Reply by piotrekkr** [reply]
Yeah they like matrix so much that they cannot fix matrix job outputs so people can actually use them in some sane way.

**Comment by olitomlinson** [comment]
Yeah this rounding-up billing model is really anti-customer. Do better.

**Comment by dataplor-mike** [comment]
5 months later and still overbilling

<img width="1059" alt="Screenshot 2023-08-23 at 12 18 19 PM" src="https://github.com/community/community/assets/113470104/df546c25-ef5a-4d9a-99d2-77702311c961">

**Comment by gtirloni** [comment]
I set up a simple workflow to check stale issues and PRs and it runs for 4 seconds. I'm billed for the full minute. Should I hide this check in some other workflow instead?

It was easy to create the workflow that I wanted that 100% solved my needs but now I'm having to re-think how I organize this because of billing issues that I shouldn't have to think about in such simple scenario, IMHO.

**Comment by alexking** [comment]
This is frustrating because by splitting up jobs to be more efficient and take less overall time, I'm actually using more billable minutes?

**Comment by doppio** [comment]
Can someone representing GitHub chime in here? Is this something that will be addressed? This is a deciding factor for me in selecting a CI/CD product.

We have a monorepo where I want to set the environment based on branch name, in the same way for all workflows, but doing so seems to require that I either:
1. Duplicate this logic across all jobs - Not desirable since it's likely to change when we add an environment
2. Select the environment in a reusable workflow, which incurs a minimum of 2 extra billable minutes for every workflow that depends on the environment: 1 billable minute for the job in the called reusable workflow, and 1 minute for the job in the calling workflow

**Reply by MatthewRyanRead** [reply]
For that specific use-case, you could create a GitHub Action rather than a reusable workflow.  Or a shell/other script might be a appropriate.

Still not totally ideal, though.

**Comment by 007mehrdad** [comment]
Uu

**Comment by EvgenyMettsler** [comment]
Up!
This issue is still unresolved and definitely needs attention. I run the cron_job which takes about 12 to 13 seconds and I billed for 1 minute each time.
So, I run this cron_job for a 100 times and was charged for a 100 minutes instead of 20 minutes, which is x5.
Common GitHub Actions team, it is almost 2024, stop rounding it, charge your customer by seconds, I'm not even asking for a milliseconds.

**Comment by jserpapinto** [comment]
We are also using more monolithic jobs because we are trying to optimize for costs instead of optimising for visibility and split of concerns.

**Comment by cheongtymm** [comment]
how this is still a thing in 2024 is beyond me.

**Comment by jaisw7** [comment]
AWS bills every customer on per-second basis. I am fairly certain that it is possible for GitHub to follow the same policy (and be profitable and maintain customer trust). 

Every non-trivial CI/CD workflow has a series of inter-dependent jobs. Every job is rounded up to a minute. 

```
Job A - 10 seconds
Job B - 11 seconds
Job C - 33 seconds
```

Total billed duration - 3 minutes. Take an objective viewpoint. Put yourself in customer's place. Your job took 1 minute. But you are being charged for 3 minutes.

**Comment by bel0v** [comment]
Seems like Github will do nothing as long as people swallow the pill and pay. I'm looking into CircleCI and other competition

**Comment by ollytheninja** [comment]
Amazing that we've made it to 2024 and this discussion is still open.
Both Azure and AWS bill their linux nodes by the second and yet good old GitHub is billing us up to 100x for job in a workflow. 
Matrix builds are a real money maker.
If runners are indeed so expensive to spin up that they need to bill for that (I doubt it) then be up front and add some fixed "start up time" rather than arbitrarily making a 10s job pay 50s "startup tax" and 50s jobs pay 10s "startup tax".

This might also encourage GitHub to build a lighter weight runner - Terraform / CDK / Serverless jobs spend a lot of time idling waiting for events to happen.

<img width="787" alt="image" src="https://github.com/community/community/assets/561757/c32c2704-985c-4d65-8f00-37c5816fc014">

**Reply by crohr** [reply]
Spawning a runner from scratch indeed takes time, about 20s in best case scenario on AWS.

I'm talking as the founder of [RunsOn](https://runs-on.com), where I've been doing extensive work to reduce queue time as much as possible when spawning ephemeral runners from scratch. So I kind of understand why GitHub would round to the nearest minute, because it IS expensive to spawn new runners.

That being said, the retail minute pricing of runners is incredibly expensive, and it should at least round to the nearest second after say 1 minute of runtime (which RunsOn is doing since AWS bills by the second).

**Comment by matthiasPOE** [comment]
Any update from anyone at github?

![Screenshot 2024-03-11 at 13 57 45](https://github.com/community/community/assets/162434935/3ac5971c-03cb-4c08-ab11-2b0f9a51c839)

**Reply by piotrekkr** [reply]
Did you run those jobs in parallel? If so then it can be displayed like this. Total time as execution time of whole workflow and each parallel job is added for billing separately which is more than total time.

**Comment by manveer-ce** [comment]
I have a dozen jobs that are just over 1 min , and rounding it up to 2mins basically 2x the costs for them..
![image](https://github.com/community/community/assets/160623539/d03d85fb-d9d2-4c5a-b41d-c469bde9a200)

**Comment by shpml** [comment]
The [docs](https://docs.github.com/en/billing/managing-billing-for-github-actions/about-billing-for-github-actions) say.

> GitHub rounds the minutes and partial minutes each job uses up to the nearest whole minute.

Run time `7m0s` billable time `8m` somehow.

![image](https://github.com/community/community/assets/74750494/07206d32-ad38-4718-ad00-ed9d3c250197)

**Reply by ianwilliams1** [reply]
Dollars to donuts that 7 minutes 0 seconds is actually some fraction of a second, which is truncated on the display, but makes it over 7 minutes, so then rounded up. Double Whammy.

**Comment by jcmv2019** [comment]
Let's also take note that aside from overbilling customers with inaccurate billable minutes, GitHub is also applying minutes multiplier to Windows and MacOS Docker machines. 

x2 for Windows and x10 for MacOS (seriously?!) It would have been understandable if MacOS' multiplier is just x5, but x10? That's way too aggressive for a service that can't accurately bill its users for the exact time it was used.

**Reply by nanderv** [reply]
Makes sense, because github runs on actual apple hardware, which requires a lot more effort to set up compared to standardised servers. Apple just doesn't support building runners, so it becomes awkward for github to do.

**Comment by mukulseth** [comment]
I just came across this, and really surprised this still hasn't been addressed in any way

**Comment by localota-travel** [comment]
Yep, we just got set up using this system and got billed a whole bunch of minutes we didn't use, for processes that took just a few seconds each. Looks like we won't be expanding on this system until this is addressed.

**Comment by Kitketovsky** [comment]
It was all easy and pleasant to setup GitHub Actions pipeline until I decided to split one job to multiple for targeting the separation of concerns and DRY, until I found out GitHub's method to bill their customers. Most likely will be looking for another CI/CD service.

**Comment by zhaoyuheng200** [comment]
I guess nobody here has came across 0 second round up to 1 minute yet, but it's true, it can happen :)

**Reply by dcmbrown** [reply]
Well.... I wrote a test action which would spin up as many runners as I specified a week or two back.  It was to definitively validate a marketplace action and also separately the service that it was using to determine which of them kept failing on an intermittent basis and specifically from GitHub.  After several tests and using up about 300 runner minutes in about 5 minutes of actual run time, I can report this is a thing. 
![Screenshot 2024-07-10 at 5 24 35 PM](https://github.com/community/community/assets/569191/696345b1-42e5-4a57-832a-0b9a83c8c10a)
In the last of those tests, which the screenshot above is a part of, I have two runners which take 0 seconds, four 1 second runners, twenty seven 2 second runners and the rest are all +3 seconds to 2m17s (for the longest).

**Comment by frankiesimon** [comment]
This really kills the idea of using this as a cheap method to run short scripts often. Too bad, I thought the setup was pretty neat.

**Reply by jpmaniqis** [reply]
> This really kills the idea of using this as a cheap method to run short scripts often. Too bad, I thought the setup was pretty neat.

**Comment by piotrekkr** [comment]
Something is quite off with how Github calculate run time of a job I have this job run:

![image](https://github.com/community/community/assets/503782/0a52d1de-b72d-403a-9948-2755a862ca6c)

It is basically 1s based on times from this screenshot. But then in summary page I have 12 seconds duration and billable time 1 minute. Also summary page shows 0 seconds for this job in dependency chart. 

![image](https://github.com/community/community/assets/503782/649be134-d0d8-4e09-926b-e5ffdc9b183b)

Is there like 11s tax added to each job run? Why three different job duration displayed? Anyone can explain this?

**Reply by MatthewRyanRead** [reply]
Time waiting for a runner, etc., is included in the overall duration.  Basically, the overall workflow is more than just its jobs.

I don't know whether that extra runtime affects the billable time though.

**Reply by piotrekkr** [reply]
GH should include this additional time somewhere in summary or billing information. How could I know how much time does it take to boot runner? Billing should be as clear as possible and displaying three different run times for same job is only confusing. Also if there can be fractions of seconds rounded up then they should also print those fractions or rounded value. Not `0s` and we pay for `1m` billable time. This is just confusing.

**Comment by mrchief** [comment]
3 years and still no recourse... I think it might be time to move to AWS Codebuild or GCP Cloudbuild or similar. As much as I love Github Actions and as much as I'm a big proponent of using them, this billing nonsense is hard to justify.

**Reply by ekwoka** [reply]
Even just billing for all of the "getting the runner setup/killing it" would be reasonable. Just arbitrarily rounding up is kind of nonsense.

It's quite useful in PR checks to have different things be unique jobs, so that you don't have one check failing preventing other checks from even running. It's nice to clearly see the specific thing that failed quickly, and also see that the other stuff is good

**Comment by almightyju** [comment]
It's insanity that this is still unchanged after all these years. 

I just split my workflow up to run some stuff in parallel to speed up the overall deployments and now I'm seeing the billable time 6-8 mins higher than the total duration, mostly because a lot of my jobs just happen to go over the minute mark by a few seconds

**Reply by SteveChurch** [reply]
I couldn't agree more. It has forced many people to create unbearable monolithic jobs and remove the use of the matrix to save them hundreds if not thousands of $$ per month. It doesn't really promote best practices being forced into doing this to avoid un-reasonable bills. 

I appreciate the comments about scale up/down however even still, the billing around these actions seem way off point. I guess if people are racking up huge bills and it's becoming a problem, that becomes the time we move to self hosted runners. 

I haven't tried [this](https://github.com/runs-on/runs-on) yet, but going to take a look. I have been playing with Blacksmith' runners too, they are nice, quite early days with supporting a wider range of runners, but rather quick in my tests and cost effective.

**Comment by aandis** [comment]
this is essentially robbery.

**Comment by lkfortuna** [comment, marked as answer]
Hi there from Actions :wave: - I understand there is a lot of community disappointment over the billing model of Actions. I understand the frustration when you see <1min of runtime in the UI and then are charged for 1 minute. I hear the feedback on being able to clearly see what is billed vs not billed, I will take that back to the team and evaluate how we can make clear the billed time for a run better. We are hoping to alleviate a bit of the pain by offering lower cost ARM runners for the community at end of year.

Just to be transparent, we have this billing model to cover the costs on our end of setting up/tearing down VMs, to keep queue times low, and to offer runners for free on public repositories. At this time we do not have plans to change to a per-second billing model. I know this will cause community frustration, but that is our reality. The way that our service works, we are not able to offer per second billing like a large scale cloud provider can, since we have to stand the VM up, install runner on the machine, install the image, etc. We are a business and we need to be able to run a net zero margin at least. I want to provide a clear answer here, so hopefully this helps to understand why we are not doing this.

**Reply by ljharb** [reply]
Is there a reason the overhead can't be billed separately from the precise time? I'd rather pay for half a second, and a separate overhead fee, than feel like I'm being cheated.

**Reply by piotrekkr** [reply]
So if my job takes 1s I pay for 59s of bootstrap time, but if my job takes 55s I pay for 5s bootstrap only. Yeah, this makes sense.

**Reply by julian-altech** [reply]
> cover the costs on our end to offer runners for free on public repositories

I had expected this to be part of it and it's probably the real bottom line reason. I still think it's dreadful to overcharge paying customers as a means to cover the free people.

**Reply by jaisw7** [reply]
> we have to stand the VM up, install runner on the machine, install the image, etc

I have the impression that GitHub uses Azure runners. Having worked at AWS and Azure, I know that **starting and tearing VM** is almost always never the case. Generally, we have a pool of worker machines (VM) which run containers (which have pre-requisites such as _**action runners**_ already installed). A user's job lands on one of the containers. Starting and stopping containers take less than 2 seconds. Therefore, rounding to a minute is frivolous. 

> we have to stand the VM up, install runner on the machine, install the image, etc

To run a simple CI job, if this is how GitHub works internally (i.e., stand/tear VM every time), please consider hiring better software engineers. As always, the open source community is always open to contributing/improving GitHub. We can help you improve the efficiency of Actions/Runners since it will help the wider developer community. Please let us know.

**Reply by piotrekkr** [reply]
There are already some services like https://github.com/runs-on/runs-on that claim that are 2x faster and 10x cheaper with 1:1 compatibility with Github runners. Afaik prices there are calculated per second so it seems like quite possible to do with proper dev and infra team.

**Reply by crohr** [reply]
@jaisw7 I think you're getting too harsh here. GHA runners are full VMs, so they take more than 2s to fully boot. Also the runners need to be kept up to date, so even if they have everything pre-baked in a machine image, that image is most likely not rebuilt nightly, so they might need to start the VM, run some update scripts, and then keep it in a warm pool so that you get the [sub-15s](https://runs-on.com/benchmarks/github-actions-runners/) starting time for your jobs. Which costs money.

@piotrekkr I'm the author of [RunsOn](https://github.com/runs-on/runs-on), thanks for the plug. RunsOn runners are indeed billed with per-second pricing (that's how AWS works), but since they take ~30s to boot, you will always get billed for those 30s (which is still 2x less than GHA runners, but just to say that there is no silver bullet here).

**Reply by jaisw7** [reply]
@crohr 

You statement is speculative in nature i.e., "**what could possibly be**". 

> GHA runners are full VMs, so they take more than 2s to fully boot

Could you please provide the source of information i.e., "**GH runners are full VMs**" and "**they take more than 2s to fully boot**"?
Containers have become mainstream in the last decade. No billion-dollar firm (such as GitHub) dedicates a full VM to run a basic CI job.

> that image is most likely not rebuilt nightly, 

A minor release of a reasonably stable piece of software such as "Github runner" happens every 7 days. So, a given VM may need to be updated (or worse restarted) every 7 days.

> so they might need to start the VM, 

No billion-dollar firm (such as GitHub) starts a full VM to run a basic CI job. A restart may happen once in 48 hours (techniques such as live VM migrations alleviate such issues).

> run some update scripts, and then keep it in a warm pool so that you get the [sub-15s](https://runs-on.com/benchmarks/github-actions-runners/) starting time for your jobs. Which costs money

Again you are speculating. Containers have become mainstream in the last decade. Sub _**milli-second**_ latency is utilized by all major cloud vendors for starting/stopping containers.

**Reply by crohr** [reply]
@jaisw7 please just read what I'm saying, and the links I've shared. Exhibit A for how the VM (yes, a VM, not a container) is built: https://github.com/actions/runner-images. Exhibit B for queuing times for various GHA providers, including GitHub: https://runs-on.com/benchmarks/github-actions-runners/

**Reply by jaisw7** [reply]
@crohr The article talks about a virtual machine image which has some packages installed (At large scale, we can't simply install operating systems and packages on every new server machine/rack the company bought yesterday - Having a VM image ensures uniformity). It **also** states that "containers" come [pre-bundled](https://github.com/actions/runner-images?tab=readme-ov-file#software-and-image-support) with VM. So, a CI job run inside a container in the virtual machine which makes sense (container's can get updated whenever Github releases a new version of action runner without updating the underlying VM. Since a minor update of a reasonably stable piece of software happens every 7 days, the containers/VM need to be updated every 7 days). The point is: You do not "start/stop" VM for every CI job (only a container which takes less than 2 seconds), and therefore the extra overhead cost (rounding to a minute) is unjustified.

**Reply by crohr** [reply]
@jaisw7 you are wrong, GHA runners are not launched within a container. And yes they are most likely cycled (i.e. destroyed & recreated) after every job, to ensure security. The link you mention is for docker support within the VM, with some base images (node, ruby, etc) preloaded on the VM, not some runner container, since the runner is a binary installed ON THE VM. I'll stop the discussion here because I don't think you are really interested in how it all works.

**Reply by jaisw7** [reply]
@crohr 

> you are wrong, GHA runners are not launched within a container

Even the third-party services such as CircleCI or self-hosted runners make use of containers. It's a pretty basic tribal knowledge. 

> yes they are most likely cycled (i.e. destroyed & recreated) after every job, to ensure security

That is again a speculation. Containers (by the very definition are containers) are secure and it is easy to create/destroy them as opposed to VMs. Fun fact: This is primarily the reason (including fast start/stop time) for proliferation of containers over VM in the current decade.

> The link you mention is for docker support within the VM, with some base images (node, ruby, etc) preloaded on the VM, not some runner container, since the runner is a binary installed ON THE VM

A docker daemon running inside a VM can pull containers (via internet). These containers have the runner binary installed (which typically get started when containers start). 

> I'll stop the discussion here because I don't think you are really interested in how it all works

Sure buddy! In the meantime, consider reading about how self-hosted runners work.

**Reply by dserodio** [reply]
@jaisw7 self-hosted runners can be safely run in containers because you're the single tenant, and you're not worried about other clients reading information from your containers. GitHub is a public service and has a free tier, and as such is subject to all kinds of abuse, like attackers running malicious GHA jobs trying to steal information from GitHub customers. VMs offer a greater level of isolation, if you search for "container escape exploit" you will find a bunch of vulnerabilities that allow these kinds of attacks to be sucessful.

**Reply by PWrzesinski** [reply]
> Hi there from Actions 👋 - I understand there is a lot of community disappointment over the billing model of Actions. I understand the frustration when you see <1min of runtime in the UI and then are charged for 1 minute. I hear the feedback on being able to clearly see what is billed vs not billed, I will take that back to the team and evaluate how we can make clear the billed time for a run better. We are hoping to alleviate a bit of the pain by offering lower cost ARM runners for the community at end of year.
> 
> Just to be transparent, we have this billing model to cover the costs on our end of setting up/tearing down VMs, to keep queue times low, and to offer runners for free on public repositories. At this time we do not have plans to change to a per-second billing model. I know this will cause community frustration, but that is our reality. The way that our service works, we are not able to offer per second billing like a large scale cloud provider can, since we have to stand the VM up, install runner on the machine, install the image, etc. We are a business and we need to be able to run a net zero margin at least. I want to provide a clear answer here, so hopefully this helps to understand why we are not doing this.

Hello, thought I'd share my opinion as well, because it seems to differ from most other responses. I came here also angry about being billed for the full minute when my quick health-check job takes like 10 seconds, but having read your explanation about the time required to set everything up I can understand it and I think it's fair. It's great that you care enough about the users to give a thorough explanation. Thanks!

**Comment by ekwoka** [comment]
@lkfortuna It's helpful to know that it is being looked at.

I think, at a minimum, having the UI do better to communicate overhead costs, at least on <1m runs could probably help reduce the feeling of being "robbed". It does make sense that there are costs to actions not directly covered by what we as users choose to run in an action, but those never being mentioned anywhere certainly makes a feeling of "what the heck is going on?"

**Comment by oleksandr-kinship** [comment]
@lkfortuna I think me same as most community can understand that GA runners have to be scaled up/down and that have to be charged. Github/Microsoft isn't charity company and you should have profit definitely, but the current billing approach looks as scam.
For example my job running 5 second. but Im charged for 1min=60sec. means 55 second you charge for instance up/down, really? I don't think that truth in 2024. Just imagining you call for Uber and instead of paying 5$ for ride you pay 60$ because your driver have to go to you some miles(but you don't know how much is it 500 meters or it's 5 miles, never mind pay 60$)
So options that could be improved: 
- make visible up/down time for instance and add these seconds to job, e.g. 5 seconds to make instance ready -> 5 seconds job running -> 3 seconds to shut down. pay for 13 seconds. (NOT 60 seconds)
- change price to include up/down charge into price. (when I buying bread in store I don't care about time and electricity that were spent to make it done, but that price included into final charge already). You should have statistics that show average start time for a server. e.g I will pay for 10 second of runtime not 0.10 but 0.11. Just an example. It could be even bargain for you because big % isn't short but you will charge more.

Anyway the model have to be changed. Charge with seconds is the right way, I want to pay for what I use, but not for something that hidden from me under hood

**Reply by nanderv** [reply]
It's a business with a pricing model. It's not a scam, because it's written down quite clearly in the docs. 

Price of a service doesn't need to correlate with actual costs; the price is based on what they think us customers are willing to pay for it. They have a service, and a pricing model. If you dislike the service/pricing model proposition, you are under no obligation to keep using this service. 
As-is, there's no *right way*, there's just the way that github/microsoft wants to monetize their service. And from their perspective I do get that I'd wanna have fewer actions runs that run for longer, instead of micro-actions.

**Reply by ben-codner-dx** [reply]
@nenderv, I understand your point, and you're not wrong about the service and pricing model being clearly outlined. However, the issue being raised here—especially with matrix jobs—creates a significant pain point for many users. When running multiple tasks concurrently, the cost quickly becomes disproportionately high due to the way the billing is structured. For example, if each task only takes a few seconds to run but you're billed for a full minute, the total cost for running a matrix can escalate rapidly, even though the actual compute time is much lower.

This also affects how pipelines are designed. For instance, if you have a pipeline with, say, 30 jobs, each running for 5 seconds, you could end up paying for 30 minutes of compute time. That kind of pricing structure forces you to rethink how you design your pipelines, often leading to unnecessary compromises just to minimize costs.

As a paid customer, it's hard not to feel frustrated when the billing doesn't align with the actual usage. I think most of us here are just looking for more transparency and fairness in the billing process.

**Reply by ekwoka** [reply]
Yeah, I think at the core, the initial "correction" is to better communicate and clarify the billing, and potentially allow sub-minute chunks. Better communicating the "per run" cost in addition to the time user code runs. This is overall the "simple" part, since it's about communication, not how the systems actually work.

Separately, a way to potentially have matrix reuse containers/runners in a smarter way, where maybe we can say "do this matrix, but only 2 at a time, and reuse them", so that the cost of create/kill is not for every single option in the matrix (maybe still with some amount of context switching cost). Or something else kind of similar. Hell, maybe something that could even better run a matrix inside a single run instance, where a step could be matrixed without matrixing the entire job (which can reuse some steps more efficiently).

While of course you could do things like that now, they don't show as separate "jobs" in many places that it might matter. So even maybe just having ways to have checks and stuff link to a specific task. A benefit I see in heavy matrix use is getting simpler to review PR checks, where you might have multiple separate things (like a multi package monorepo) that you'd want separate checks for.

**Comment by jeremyjpj0916** [comment]
2026 and nothing changed lol. GG.

**Reply by Animalmix55** [reply]
The house always wins!
